js/data/feature_tree.js

```diff
@@ -88,80 +152,181 @@ FeatureTree.prototype.query = function(args, styleLayersByID) {
             }
-
-            result.push(util.extend({layer: layerID}, geoJSON));
         }
     }
+};
 
-    function translate(translate, translateAnchor) {
-        if (!translate[0] && !translate[1]) {
-            return queryGeometry;
-        }
+FeatureTree.prototype.makeGeoJSON = function(featureIndexArray, styleLayers) {
+    if (!this.vtLayers) {
+        this.vtLayers = new vt.VectorTile(new Protobuf(new Uint8Array(this.rawTileData))).layers;
+        this.sourceLayerNumberMapping = new StringNumberMapping(this.vtLayers ? Object.keys(this.vtLayers).sort() : ['_geojsonTileLayer']);
+    }
 
-        translate = Point.convert(translate);
+    var result = {};
 
-        if (translateAnchor === "viewport") {
-            translate._rotate(-args.bearing);
+    featureIndexArray = new FilteredFeatureIndexArray(featureIndexArray);
+    var indexes = featureIndexArray.get(0);
+
+    var cachedLayerFeatures = {};
+    for (var i = 0; i < featureIndexArray.length; i++) {
+        indexes._setIndex(i);
+        var sourceLayerName = this.sourceLayerNumberMapping.numberToString[indexes.sourceLayerIndex];
+        var sourceLayer = this.vtLayers[sourceLayerName];
+        var featureIndex = indexes.featureIndex;
+
+        var cachedFeatures = cachedLayerFeatures[sourceLayerName];
+        if (cachedFeatures === undefined) {
+            cachedFeatures = cachedLayerFeatures[sourceLayerName] = {};
         }
 
-        var translated = [];
-        for (var i = 0; i < queryGeometry.length; i++) {
-            translated.push(queryGeometry[i].sub(translate._mult(pixelsToTileUnits)));
+        var feature = cachedFeatures[featureIndex] = cachedFeatures[featureIndex] || sourceLayer.feature(featureIndex);
+
+        var layerID = this.numberToLayerIDs[indexes.bucketIndex][indexes.layerIndex];
+        var styleLayer = styleLayers[layerID];
+        if (!styleLayer) continue;
+
+        var geojsonFeature = new GeoJSONFeature(feature, this.z, this.x, this.y);
+        geojsonFeature.layer = styleLayer.serialize({
+            includeRefProperties: true
+        });
+        var layerResult = result[layerID];
+        if (layerResult === undefined) {
+            layerResult = result[layerID] = [];
         }
-        return translated;
+        layerResult.push(geojsonFeature);
     }
 
     return result;
 };
 
+function translate(queryGeometry, translate, translateAnchor, bearing, pixelsToTileUnits) {
+    if (!translate[0] && !translate[1]) {
+        return queryGeometry;
+    }
+
+    translate = Point.convert(translate);
+
+    if (translateAnchor === "viewport") {
+        translate._rotate(-bearing);
+    }
+
+    var translated = [];
+    for (var i = 0; i < queryGeometry.length; i++) {
+        translated.push(queryGeometry[i].sub(translate._mult(pixelsToTileUnits)));
+    }
+    return translated;
+}
+
+function matchLayers(filterLayerIDs, featureLayerIDs) {
```

Comment:
Can we replace this with a generic `util` method? Something like `util.intersection`